components/tidb_query_expr/src/impl_regexp.rs

```diff
@@ -0,0 +1,1322 @@
+    let pattern = match args[PATTERN_IDX].as_bytes() {
+        Some(b) => b,
+        None => return Ok(None),
+    };
+
+    let match_type = if args.len() > match_idx {
+        match args[match_idx].as_bytes() {
+            Some(b) => b,
+            None => return Ok(None),
+        }
+    } else {
+        b""
+    };
+
+    build_regexp::<C>(pattern, match_type).map(|reg| Some(reg))
+}
+
+fn init_regexp_data<C: Collator, const N: usize>(expr: &mut Expr) -> Result<Option<Regex>> {
+    let children = expr.mut_children();
+    if children.len() <= PATTERN_IDX {
+        return Ok(None);
+    }
+
+    let pattern = match children[PATTERN_IDX].get_tp() {
+        ExprType::Bytes | ExprType::String => children[PATTERN_IDX].get_val(),
+        _ => return Ok(None),
+    };
+
+    let match_type = if children.len() > N {
+        match children[N].get_tp() {
+            ExprType::Bytes | ExprType::String => children[N].get_val(),
+            _ => return Ok(None),
+        }
+    } else {
+        b""
+    };
+
+    build_regexp::<C>(pattern, match_type).map(|reg| Some(reg))
+}
+
+/// Currently, TiDB only supports regular expressions for utf-8 strings.
+/// See https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-like.
+#[rpn_fn(nullable, raw_varg, min_args = 2, max_args = 3, capture = [metadata], metadata_mapper = init_regexp_data::<C, LIKE_MATCH_IDX>)]
+#[inline]
+pub fn regexp_like<C: Collator>(
+    metadata: &Option<Regex>,
+    args: &[ScalarValueRef<'_>],
+) -> Result<Option<i64>> {
+    let expr = match args[0].as_bytes() {
+        Some(e) => std::str::from_utf8(e)?,
+        None => return Ok(None),
+    };
+    let regex = match metadata {
+        Some(r) => r.clone(),
+        None => match build_regexp_from_args::<C>(args, LIKE_MATCH_IDX)? {
+            Some(r) => r,
+            None => return Ok(None),
+        },
+    };
+
+    Ok(Some(regex.is_match(expr) as i64))
+}
+
+/// Currently, TiDB only supports regular expressions for utf-8 strings.
+/// See https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-substr.
+#[rpn_fn(nullable, raw_varg, min_args = 2, max_args = 5, capture = [metadata], metadata_mapper = init_regexp_data::<C, SUBSTR_MATCH_IDX>)]
+#[inline]
+pub fn regexp_substr<C: Collator>(
+    metadata: &Option<Regex>,
+    args: &[ScalarValueRef<'_>],
+) -> Result<Option<Bytes>> {
+    let mut expr = match args[0].as_bytes() {
+        Some(e) => std::str::from_utf8(e)?,
+        None => return Ok(None),
+    };
+    let regex = match metadata {
+        Some(r) => r.clone(),
+        None => match build_regexp_from_args::<C>(args, SUBSTR_MATCH_IDX)? {
+            Some(r) => r,
+            None => return Ok(None),
+        },
+    };
+
+    if args.len() >= 3 {
+        let pos = match EvaluableRef::borrow_scalar_value_ref(args[2]) {
+            Some::<&i64>(p) => *p,
+            None => return Ok(None),
+        };
+
+        let count = expr.chars().count() as i64;
```

Comment:
`chars().count()` also iterates over the whole string. I mean we can check `pos >= 1` first, then if ` expr.char_indices().nth((pos - 1) as usize)` returns `None`, we can also return error.

Reply:
Got it.

Reply:
Done.